### PR TITLE
Fetching game servers from multiple hubs

### PR DESCRIPTION
**Pull request by Visne**
Adds a "Hub Settings" option where the user can set what hubs to fetch game servers from.

Minor things this PR also does:
- Add left margin to scroll bar
- Fix padding for OpenBoth/OpenRight/OpenLeft style classes
- Use custom SS14 logo in all windows instead of just main window

Game server descriptions don't appear to work right now, but I'm 99% sure that's a hub issue, not a launcher issue

### CODE REVIEW

File: SS14.Launcher/Api/HubApi.cs
Comment (PJB3005) [inline SS14.Launcher/Api/HubApi.cs:99]: Wouldn't it make more sense to just not have the try catch in that case?

File: SS14.Launcher/Api/HubApi.cs
Comment (PJB3005) [inline SS14.Launcher/Api/HubApi.cs:86]: This is O(n²). Keep a hash set around for this function instead.

File: SS14.Launcher/HappyEyeballsHttp.cs
Comment (PJB3005) [inline SS14.Launcher/HappyEyeballsHttp.cs:24]: It's probably better to set timeouts locally via `CancellationToken`. We do this for the server status fetching logic.

File: SS14.Launcher/Api/HubApi.cs
Comment (PJB3005) [inline SS14.Launcher/Api/HubApi.cs:106]: Same thing here. This should take two URIs rather than being knowledgeable of `ServerStatusData`.

File: SS14.Launcher/Models/Data/DataManager.cs
Comment (PJB3005) [inline SS14.Launcher/Models/Data/DataManager.cs:595]: Address is the primary key, you shouldn't be needing to filter for `Priority` in the `WHERE`.

File: SS14.Launcher/Models/Data/Migrations/Script0006_Hubs.sql
Comment (PJB3005) [inline SS14.Launcher/Models/Data/Migrations/Script0006_Hubs.sql:1]: There's some fancy relational algebra reason for this, but table names shouldn't be plural (which the rest of the launcher tables also follow).

File: SS14.Launcher/Views/HubSettingsDialog.xaml.cs
Comment (PJB3005) [inline SS14.Launcher/Views/HubSettingsDialog.xaml.cs:44]: Move this to the view model.

**Review by PJB3005 — CHANGES_REQUESTED** [review]
The default list of hubs shouldn't be removable from the launcher. 

Reasons:
* Avoids people screwing up their launcher and not knowing why it doesn't list servers. (Yes I see there's a reset button, not good enough).
* Makes testing against local hubs easier.
* I'm still planning to change the URL some time in the future and I'd prefer to be able to change it properly rather than having to keep the old URLs in an nginx config somewhere permanently.
* Don't need to copy it into the migration file (this one is fixable but ehhhh)

**Comment by Visne** [comment]
Code reviews are resolved.

I'm strongly opposed however to locking the default hubs in the UI. Firstly just on principle, because I believe players should have the freedom to just pick whatever hubs they prefer. Secondly, I think that if players want such a feature in the future, they will fork the code and just distribute their own launcher. That defeats the entire purpose of this PR (or at least, the reason I started coding it) which is to avoid people from doing that and being able to easily spread malware. I understand your concerns however, but I think they can be fixed:

> Avoids people screwing up their launcher and not knowing why it doesn't list servers. (Yes I see there's a reset button, not good enough).

It's probably a good idea to make it impossible to set 0 hubs, since I can't think of a valid reason to do that. Since the UI will stop you from setting invalid URI's, that means a player has to remove the official hub, and then somehow accidentally write a valid URI and then save it. Furthermore, I can add a big reset button to the error message in the server list, which shows up when no servers are loaded. That way, if a player still manages to mess up their hub settings, they are intuitively led to a button that fixes the problem right away.

> Makes testing against local hubs easier.

In my opinion it only makes it harder. In fact, my current hub settings don't have the Space Wizards hub as I only added my local one, to quickly see if the actual servers on my local hub show up. I could have changed the hard coded hub URL, but obviously just using the normal UI is much simpler, and could maybe avoid introducing weird edge-cases.

> I'm still planning to change the URL some time in the future and I'd prefer to be able to change it properly rather than having to keep the old URLs in an nginx config somewhere permanently.

I don't really understand how this relates to locking the default hubs list. A migration can just replace the old hub URL with the new one if it exists, and if not it can just do nothing.

> Don't need to copy it into the migration file (this one is fixable but ehhhh)

I don't really see why this is a problem, but I would be glad to spend time to fix it.

**Comment by metalgearsloth** [comment]
> The default list of hubs shouldn't be removable from the launcher.

Is non-negotiable so if someone else picks that up that still needs doing.

**Comment by PrPleGoo** [comment]
> > The default list of hubs shouldn't be removable from the launcher.
> 
> Is non-negotiable so if someone else picks that up that still needs doing.

image

So "additional" instead of "alternative" was meant here?

**Comment by metalgearsloth** [comment]
> So "additional" instead of "alternative" was meant here?

image

**Comment by Crossedfall** [comment]
I don't think the image does as much as you think it does. It's clear and obvious that most people understood the statement to mean that people could choose to use unofficial options **instead of** the official hub. Which is the direct noun definition of "alternative."

I think the author provided valid counterpoints on how to address the problems presented. Closing the pr and acting needlessly dismissive doesn't really help. Clarifying the orgs position on why it's, "non-negotiable" would have been a significant step up from the actual response that was provided.

**Comment by Skyedra** [comment]
This is pretty cool @Visne :+1: 

IMO the inclusion of the reset feature mitigates the concern of a user borking their hub settings as it seems fairly clearly labeled and easy to do:

https://github.com/space-wizards/SS14.Launcher/assets/22365940/0d7c4dcb-7f5d-40f2-8646-3d33bc8d140e

**Comment by PrPleGoo** [comment]
If I do not want to see servers from the default hub, what alternative do I have? The servers from the default hub are always forced on me, even if I am seeking alt hubs to e.g. avoid the steam tide.

**Comment by Visne** [comment]
There is some context missing from this PR that make things a bit unclear.

On the 13th of June (so after my previous comment but before the PR was closed) I was approached in a private maintainer chat in the SS14 Discord, and told that I would have to code what PJB requested for the PR to be merged. I declined, for the same reasons as I have stated in my previous comment. Then a few days later the PR was closed.

 If any SS14 maintainer/project manager feels that any part of that is inaccurate then please let me know and I will try to correct it.

And regarding the code, I consider it to be MIT licensed (and I think it would be legally anyways) so anyone that wants to implement the feature PJB requested can do so.

**Comment by PJB3005** [comment]
Alright I'm once again going PR-mask-off mode to give you our full stance on this PR:

The reason we don't like this PR is simple: it enables a group of [serial harassers](https://forum.spacestation14.io/index.php?/topic/1951-clarification-towards-nyanotrasen-our-handling-to-the-situation-and-changes-from-here-on/) to fracture the community.

In principle, I do agree that giving people more technical control is good. However, technology is not apolotical, and I know for a fact that this feature will be used for bad by the people mentioned above.

These people have **not stopped harassing us, continued causing serious harm to the project**, and even now are continuing to astroturf this PR. Quite frankly, I do not want to give them any concessions on the official launcher.

*If you really insist about twisting my arm*, we've decided you can have the following: instead of being able to remove the official hub, you can toggle individual hubs in the filters menu of the server browser. So you'd be able to hide the official hub through that if you want.

**Comment by DocNITE** [comment]
Well, it's really important feature...

**Comment by Visne** [comment]
> you can toggle individual hubs in the filters menu of the server browser

This is a compromise I was thinking of as well, and I think it's a very good idea.

**Comment by Cheackraze** [comment]
> astroturf this PR.

Either you wildly underestimate the support for this or you don't understand what astroturfing means.  Your resistance to federation for community moderation is harming the communities's reputation and scaring off potential communities from adopting the platform.

**Comment by ike709** [comment]
> Alright I'm once again going PR-mask-off mode to give you our full stance on this PR:

I think you underestimate how much damage your pissing match with Nyano is doing to the project's reputation in the eyes of people who aren't all aboard the SS14 train already.

**Comment by PowerfulBacon** [comment]
> Alright I'm once again going PR-mask-off mode to give you our full stance on this PR:

I appreciate the honesty, its the answer we was all expected. I do wonder, could allowing them to split off and have their own hidden community be beneficial for SS14? It would allow the communities that have been causing struggles for SS14 as a whole to form their own independant communities where there problems become more their own, and since they are hidden away the serious reputation hits that SS14 takes from these incidents would be reduced as they won't be related to the official project management anymore. It would let SS14 take their own direction as a server while still allowing communities that have been oppossed to SS14's leadership decisions to go their own way rather than have a reason to harass and drive away the developers for the project.

Ultimately, SS14 needs a core codebase with minimal drama in order to have more positive communities sprout from it, if the core codebase is known for drama and fighting against its downstreams then any codebases that may bring positive movement for the project may be driven away and you will only be entering a death spiral. Driving away the negative communities needs to be subtle process with minimal risk of dividing a large number of people's opinions otherwise you will get fracturing whether or not players can choose to abandon the official hub anyway.

In the end do you really want the people who don't want to play on the official hub playing on them? If they are strongly against the core management, forcing them through you may only add fuel to the fire.

Just my two cents, I'm sure this will get figured out eventually and the drama will eventually fade away as long as more drama doesn't continue to get created.

Also, from a feature perspective, what's the difference between someone removing the primary server and having it set to always be hidden if the reset button will always bring it back and has the official servers hardcoded? Isn't it functionally identical to the implementation in this PR at that point? If the difference is that it automatically unfilters on load, then the only thing changed is the removal of a QOL of feature as players wishing to avoid the official servers will always set that hide flag anyway and there are no functional changes.

**Comment by moonheart08** [comment]
the launcher has a direct connect button and favorites list if they want their own community.

**Comment by PJB3005** [comment]
I do agree that letting people go their own way would be the lowest friction option, which is why we agreed that alt hubs should be supported in the first place. The distinction of "removing main hub not allowed", and why we think the filters thing is better instead is somewhat more nuanced:

We don't want this feature getting used to insulate players away from the main hub. If somebody hard-replaces the main hub for a smaller hub, that puts them far further away and more likely to never engage with the main hub again. Sure, there is a reset button, but that's annoying and it's clear the idea is more a "set and forget" than something you frequently interact with. Because of this we also don't consider this feature very suitable for smaller curated subs, like the "avoid steamtide" example given. I don't think people who want to avoid steamtide want to necessarily make it harder to see the original list.

We also just really want to prevent the "SS14 is great but the main hub is ran by power freaks, you should immediately swap it out with The Good One™️"-type discourse. That's the fracturing we want to avoid. 

The filters approach solves this all far better: players can pick their own hub but still easily have the option to browse between all. 

Why filters weren't thought of earlier? 🤷‍♀️ part of it is just that the alt hubs idea was already formulated in our head by the time filters were implemented. Obvious in hindsight.

**Comment by SelfDrivingRiddler** [comment]
> We also just really want to prevent the "SS14 is great but the main hub is ran by power freaks, you should immediately swap it out with The Good One™️"-type discourse. That's the fracturing we want to avoid.

**Comment by Visne** [comment]
> We don't want this feature getting used to insulate players away from the main hub. If somebody hard-replaces the main hub for a smaller hub, that puts them far further away and more likely to never engage with the main hub again. Sure, there is a reset button, but that's annoying and it's clear the idea is more a "set and forget" than something you frequently interact with.

That sounds very reasonable and understandable, and clears up some confusion I still had.

There's one thing I'm still wondering though; when the hub filters are implemented, would it still be unreasonable to have the ability to remove the main hub? The only practical difference would be that if you really want to get rid of it for some reason, you don't have to keep staring at a filter that is permantenly disabled. You could probably argue that it would still be possible for people to accidentally remove it, but I would be fine with adding a warning if someone tries to remove the default hub (with a big cancel button that undoes the changes).

I think symbolically it would also be a nice, there's a big difference between having a default and having a default that is impossible to remove.

More directly though, I would much prefer spending my time coding a little warning than implementing the logic to make it so that the default hub can't be deleted, since I think it's more of an anti-feature anyways.

Oh and by the way, I will add some code to make it impossible to have no hubs, which should make it even harder to mess up the hub config.

**Comment by Skyedra** [comment]
Hi @Visne,

I had two requests for consideration -- perhaps you are already planning these as I vaguely remember them being discussed but couldn't re-find the discussions so figured I'd best post here:

1) When trying this code with a fresh profile (via the `SS14_LAUNCHER_APPDATA_NAME=someNewValueHere` environment variable), it seems no hubs are prepopulated.  Would you be willing / or maybe already planning to prepopulate the hubs with everything in DefaultHubUrls?  Then users won't have to manually reset their hubs in settings on first install.

2) Ideally, if two hubs list the same server, the entries are combined instead of showing duplicates:

image

Servers may want to list in multiple hubs and those hubs may not be exactly in sync with each other.  Perhaps the topmost hub in the settings panel should take priority if two entries matching addresses are found?

(For testing purposes, you can add `https://cdn.blepstation.com/hub/` to try this -- note that URL only mirrors the main server list on a time delay and doesn't attempt to mirror each of the 78 server requests, which in this case might be useful for seeing if there are duplicates or not.)

**Comment by Visne** [comment]
Thanks @Skyedra, that's very useful.

Point 1 is definitely a bug and I will fix it as soon as I have time.

Point 2 is also a bug, there's already code that should handle that but I think I broke it will refactoring. Currently it checks for duplicates based on server IP, hub URI and status data. Obviously it should only check for duplicates based on server IP.

**Comment by PixelTheKermit** [comment]
Not sure if it's coded or not, but it would be cool to see which server is from which hub

**Comment by Visne** [comment]
It would be a nice feature, but I don't plan on adding it in this PR.

**Comment by Visne** [comment]
> When trying this code with a fresh profile (via the SS14_LAUNCHER_APPDATA_NAME=someNewValueHere environment variable), it seems no hubs are prepopulated. Would you be willing / or maybe already planning to prepopulate the hubs with everything in DefaultHubUrls? Then users won't have to manually reset their hubs in settings on first install.

So it turns out that this isn't actually an issue with this PR. If you were to restart the launcher after doing that (without touching the hub settings at all) you can see that this time the servers are loaded fine. See #95 for more details. It should definitely be fixed before this PR is merged.

**Comment by MrStonedOne** [comment]
If this pr is not merged with the ability to disable the default hub, ***I*** will fractor the community and fully fork ss14 when tg goes to it.

engine, launcher, everything.

**Comment by Skyedra** [comment]
As someone with a (small/wip) launcher fork, I'd kindly point out that it is healthy for there to be multiple forks of the launcher.  Then people can choose their favorite launcher based on their use-case/needs.  :)

The real problem is if launchers are not compatible with each other's servers.  One of the reasons this may happen would be due to differing account systems.  Hopefully we can come up with a good way for launcher+engine to support multi-account as this would alleviate a lot of future incompatibility related problems IMO.

**Comment by metalgearsloth** [comment]
We have a discord https://discord.gg/6tHc6pfG

**Comment by Visne** [comment]
Now that #95 is fixed, all the bugs that @Skyedra pointed out are also fixed. It's now ready for review.

Note that the current code still allows for removing the default hub, although it now forbids you from setting zero hubs. I believe it makes the most sense to do it that way, and I would still like a formal response to the reasons stated in https://github.com/space-wizards/SS14.Launcher/pull/87#issuecomment-1619408075.

**Comment by PJB3005** [comment]
I still insist it should not allow removing the default hub.

**Comment by Jezithyr** [comment]
It should also make it clear which servers are not from the main hub

**Comment by Visne** [comment]
> I still insist it should not allow removing the default hub.

Alright, I'm on it.

> It should also make it clear which servers are not from the main hub

That's something @PixelTheKermit also suggested and I think it's a good idea, but it's a bit outside of the scope of this PR.

**Comment by Visne** [comment]
Done

**Comment by Jezithyr** [comment]
> > I still insist it should not allow removing the default hub.
> 
> Alright, I'm on it.
> 
> > It should also make it clear which servers are not from the main hub
> 
> That's something @PixelTheKermit also suggested and I think it's a good idea, but it's a bit outside of the scope of this PR.

It's not really, we should be clear that these servers are not officially sanctioned

**Comment by Visne** [comment]
Alright, done.

File: SS14.Launcher/ViewModels/HubSettingsViewModel.cs
Comment (PJB3005) [inline SS14.Launcher/ViewModels/HubSettingsViewModel.cs:29]: Wouldn't it make more sense to do this check *before* running it through `new Uri()`?

File: SS14.Launcher/Views/HubSettingsDialog.xaml
Comment (PJB3005) [inline SS14.Launcher/Views/HubSettingsDialog.xaml:117]: This probably shouldn't be a multi-line textbox?

**Review by PJB3005 — CHANGES_REQUESTED** [review]
The styling changes seem to have broken the layout for the player count filters: 
image

File: SS14.Launcher/ViewModels/HubSettingsViewModel.cs
Comment (Visne) [reply SS14.Launcher/ViewModels/HubSettingsViewModel.cs:32]: Should I just modify `IsValidHubUri()` to reject anything that has query params/fragments/etc in them?

File: SS14.Launcher/Models/Data/Migrations/Script0006_Hubs.sql
Comment (Visne) [reply SS14.Launcher/Models/Data/Migrations/Script0006_Hubs.sql:12]: Why not? Doing it differently would require special code for populating the UI and storing the hubs to the DB, while the current approach is simple and uniform

File: SS14.Launcher/Views/HubSettingsDialog.xaml
Comment (Visne) [reply SS14.Launcher/Views/HubSettingsDialog.xaml:117]: Are you sure it's multiline?

Comment (PJB3005) [reply SS14.Launcher/Views/HubSettingsDialog.xaml:117]: Yes it is, paste some multi-line text into it.

File: SS14.Launcher/ViewModels/HubSettingsViewModel.cs
Comment (PJB3005) [reply SS14.Launcher/ViewModels/HubSettingsViewModel.cs:32]: Probably.

File: SS14.Launcher/Models/Data/Migrations/Script0006_Hubs.sql
Comment (PJB3005) [reply SS14.Launcher/Models/Data/Migrations/Script0006_Hubs.sql:12]: Conceptually it seems easier to special case default hubs if they *aren't* in the database. You're currently storing default hubs in the DB and then need to do annoying logic to detect if it's a default hub, vs if the data was stored separately and as such wouldn't need splitting apart in the first place.

File: SS14.Launcher/Views/HubSettingsDialog.xaml
Comment (Visne) [reply SS14.Launcher/Views/HubSettingsDialog.xaml]: guh

File: SS14.Launcher/ViewModels/HubSettingsViewModel.cs
Comment (Visne) [reply SS14.Launcher/ViewModels/HubSettingsViewModel.cs:29]: I figured this was more robust, specifically since it would trim whitespace but maybe there's other edge cases I haven't thought of. And since it's only done when the user presses OK in the menu, I imagine it doesn't matter for performance in practice?

File: SS14.Launcher/Views/HubSettingsDialog.xaml
Comment (Visne) [reply SS14.Launcher/Views/HubSettingsDialog.xaml]: Since this also affects all other TextBoxes in the launcher, and I have no idea how to fix it, can I resolve this for now? Doesn't seem like blocking issue to me, a multiline string won't be a valid URI anyways unless the other lines are whitespace.

File: SS14.Launcher/Models/Data/Migrations/Script0006_Hubs.sql
Comment (Visne) [reply SS14.Launcher/Models/Data/Migrations/Script0006_Hubs.sql:12]: Currently the 'annoying logic' is just a single `Contains()`. If I didn't store the default hubs in the DB, then I would have to separately add the default hubs to populate the UI window, and then when storing in the DB I would have to do the `Contains()` call regardless, to make sure I don't accidentally add the default hub to the DB. Seems much more annoying and less robust to me.

Comment (PJB3005) [reply SS14.Launcher/Models/Data/Migrations/Script0006_Hubs.sql:12]: If properly structured, those lists would be separated for UI display purposes and would never need to be separated.

File: SS14.Launcher/Views/HubSettingsDialog.xaml
Comment (PJB3005) [reply SS14.Launcher/Views/HubSettingsDialog.xaml]: Uhhh yeah I guess.

File: SS14.Launcher/ViewModels/HubSettingsViewModel.cs
Comment (PJB3005) [reply SS14.Launcher/ViewModels/HubSettingsViewModel.cs:29]: Meh it's minor so I guess do whatever.

File: SS14.Launcher/Models/Data/Migrations/Script0006_Hubs.sql
Comment (Visne) [reply SS14.Launcher/Models/Data/Migrations/Script0006_Hubs.sql:12]: See Discord: https://discord.com/channels/310555209753690112/770682801607278632/1155937079482667090

**Review by PJB3005 — APPROVED** [review]

**Pull request merged**
